typescript/agentkit/src/wallet-providers/privyWalletProvider.ts

```diff
@@ -0,0 +1,165 @@
+import { PrivyClient } from "@privy-io/server-auth";
+import { createViemAccount } from "@privy-io/server-auth/viem";
+import { ViemWalletProvider } from "./viemWalletProvider";
+import { createWalletClient, http, WalletClient } from "viem";
+
+import * as chains from "viem/chains";
+
+/**
+ * Get a chain from the viem chains object
+ *
+ * @param id - The chain ID
+ * @returns The chain
+ */
+function getChain(id: number) {
+  const chainList = Object.values(chains);
+  return chainList.find(chain => chain.id === id);
+}
+
+/**
+ * Configuration options for the Privy wallet provider.
+ *
+ * @interface
+ */
+interface PrivyWalletConfig {
+  /** The Privy application ID */
+  appId: string;
+  /** The Privy application secret */
+  appSecret: string;
+  /** The ID of the wallet to use, if not provided a new wallet will be created */
+  walletId?: string;
+  /** Optional network ID to connect to */
+  chainId?: number;
+  /** Optional authorization key for the wallet API */
+  authorizationPrivateKey?: string;
+  /** Optional authorization key ID for creating new wallets */
+  authorizationKeyId?: string;
+}
+
+/**
+ * A wallet provider that uses Privy's server wallet API.
+ * This provider extends the ViemWalletProvider to provide Privy-specific wallet functionality
+ * while maintaining compatibility with the base wallet provider interface.
+ */
+export class PrivyWalletProvider extends ViemWalletProvider {
+  #walletId: string;
+  #authorizationPrivateKey: string | undefined;
+
+  /**
+   * Private constructor to enforce use of factory method.
+   *
+   * @param walletClient - The Viem wallet client instance
+   * @param config - The configuration options for the Privy wallet
+   */
+  private constructor(
+    walletClient: WalletClient,
+    config: PrivyWalletConfig & { walletId: string }, // Require walletId in constructor
+  ) {
+    super(walletClient);
+    this.#walletId = config.walletId; // Now guaranteed to exist
+    this.#authorizationPrivateKey = config.authorizationPrivateKey;
+  }
+
+  /**
+   * Creates and configures a new PrivyWalletProvider instance.
+   *
+   * @param config - The configuration options for the Privy wallet
+   * @returns A configured PrivyWalletProvider instance
+   *
+   * @example
+   * ```typescript
+   * const provider = await PrivyWalletProvider.configureWithWallet({
+   *   appId: "your-app-id",
+   *   appSecret: "your-app-secret",
+   *   walletId: "wallet-id",
+   *   networkId: "base-sepolia"
+   * });
+   * ```
+   */
+  public static async configureWithWallet(config: PrivyWalletConfig): Promise<PrivyWalletProvider> {
+    const privy = new PrivyClient(config.appId, config.appSecret, {
```

Comment:
There is an edge case here where the `authorizationPrivateKey` is provided but not `authorizationKeyId` which will get things into a funky state.  Can you do a check to enforce that either both or neither are provided before initializing the `PrivyClient`?

Reply:
I think it's ok to initialise with just an `authorizationPrivateKey` if you're using an existing wallet, not creating a key?

Reply:
Ah, you're right - if the user has an existing wallet that has a key attached to it, you need to specify `authorizationPrivateKey` and technically don't need to specify the ID.  My bad! Fine to leave as is, good catch.